File: sdk/ruby/test_standalone.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Standalone test runner — no bundler or rspec required.
# Exercises core SDK functionality using stdlib only (no HTTP).
# Run: ruby test_standalone.rb

$LOAD_PATH.unshift File.join(__dir__, 'lib')
require 'experimentation_platform'

failures = []

def assert_equal(description, expected, actual, failures)
  if (actual - expected).abs < 1e-10
    puts "  PASS: #{description}"
  else
    msg = "FAIL: #{description}\n       expected=#{expected}, got=#{actual}, delta=#{(actual - expected).abs}"
    puts "  #{msg}"
    failures << msg
  end
end

def assert_nil(description, actual, failures)
  if actual.nil?
    puts "  PASS: #{description}"
  else
    msg = "FAIL: #{description} — expected nil, got #{actual.inspect}"
    puts "  #{msg}"
    failures << msg
  end
end

def assert_true(description, actual, failures)
  if actual
    puts "  PASS: #{description}"
  else
    msg = "FAIL: #{description} — expected true, got #{actual.inspect}"
    puts "  #{msg}"
    failures << msg
  end
end

def assert_in_range(description, value, min, max, failures)
  if value >= min && value < max
    puts "  PASS: #{description}"
  else
    msg = "FAIL: #{description} — #{value} not in [#{min}, #{max})"
    puts "  #{msg}"
    failures << msg
  end
end

puts "=" * 60
puts "ExperimentationPlatform Ruby SDK — Standalone Test Runner"
puts "=" * 60
puts

# ---------------------------------------------------------------------------
# Test group 1: Hash parity (CRITICAL cross-SDK test)
# ---------------------------------------------------------------------------
puts ">>> Hash Parity Tests"

result   = ExperimentationPlatform::FeatureFlagEvaluator.hash_user("user-123", "my-flag")
expected = 0.6927449859213084
assert_equal(
  'hash_user("user-123", "my-flag") == 0.6927449859213084  [cross-SDK parity]',
  expected, result, failures
)

result2 = ExperimentationPlatform::FeatureFlagEvaluator.hash_user("user-123", "my-flag")
assert_equal('Determinism: same call twice yields identical result', result, result2, failures)

v1 = ExperimentationPlatform::FeatureFlagEvaluator.hash_user("user-456", "my-flag")
assert_in_range('hash_user("user-456", "my-flag") is in [0.0, 1.0)', v1, 0.0, 1.0, failures)

v_empty = ExperimentationPlatform::FeatureFlagEvaluator.hash_user("", "")
assert_in_range('hash_user("", "") is in [0.0, 1.0)', v_empty, 0.0, 1.0, failures)

assert_true('Local evaluation was removed (server decides)',
  !ExperimentationPlatform::FeatureFlagEvaluator.respond_to?(:evaluate), failures)

puts

# ---------------------------------------------------------------------------
# Test group 2: Result types
# ---------------------------------------------------------------------------
puts ">>> Result Type Tests"

flag = ExperimentationPlatform::FlagEvaluation.new(key: 'dark-mode', enabled: true, config: { 'theme' => 'dark' })
assert_true('FlagEvaluation#enabled? reflects enabled', flag.enabled? == true, failures)
assert_true('FlagEvaluation#to_h exposes key/enabled/config',
  flag.to_h == { key: 'dark-mode', enabled: true, config: { 'theme' => 'dark' } }, failures)

assignment = ExperimentationPlatform::Assignment.new(
  experiment_key: 'exp', variant_id: 'v1', variant_name: 'control', is_control: true, configuration: nil
)
assert_true('Assignment#control? reflects is_control', assignment.control? == true, failures)
assert_true('Assignment exposes variant_name', assignment.variant_name == 'control', failures)

batch = ExperimentationPlatform::BatchResult.new(success_count: 2, failure_count: 0, errors: nil)
assert_true('BatchResult#ok? is true without failures', batch.ok?, failures)

puts

# ---------------------------------------------------------------------------
# Test group 3: Cache
# ---------------------------------------------------------------------------
puts ">>> Cache Tests"

cache = ExperimentationPlatform::Cache.new(ttl: 60, max_size: 3)
assert_nil('Cache: get returns nil for missing key', cache.get('x'), failures)
cache.set('x', 42)
val = cache.get('x')
assert_true('Cache: get returns stored value', val == 42, failures)
cache.clear
assert_nil('Cache: clear removes all entries', cache.get('x'), failures)

# Eviction test
3.times { |i| cache.set("k#{i}", i) }
cache.set('k3', 99)  # should evict k0
evicted = cache.get('k0')
assert_nil('Cache: evicts oldest when at max_size', evicted, failures)
kept = cache.get('k3')
assert_true('Cache: new entry kept after eviction', kept == 99, failures)

# Per-user listing
cache.clear
cache.set([:flag, 'u1', 'a'], 'A')
cache.set([:flag, 'u2', 'b'], 'B')
listed = cache.live_values { |key| key[1] == 'u1' }
assert_true('Cache: live_values lists entries for one user', listed == ['A'], failures)

puts

# ---------------------------------------------------------------------------
# Test group 4: Errors hierarchy
# ---------------------------------------------------------------------------
puts ">>> Error Hierarchy Tests"

assert_true('NetworkError is-a Error',
  ExperimentationPlatform::NetworkError.ancestors.include?(ExperimentationPlatform::Error),
  failures)
assert_true('APIError is-a Error',
  ExperimentationPlatform::APIError.ancestors.include?(ExperimentationPlatform::Error),
  failures)
assert_true('AuthenticationError is-a APIError',
  ExperimentationPlatform::AuthenticationError.ancestors.include?(ExperimentationPlatform::APIError),
  failures)

err = ExperimentationPlatform::APIError.new("bad request", status_code: 400)
assert_true('APIError#status_code returns 400', err.status_code == 400, failures)

puts

# ---------------------------------------------------------------------------
# Test group 5: Config validation
# ---------------------------------------------------------------------------
puts ">>> Config Validation Tests"

begin
  ExperimentationPlatform::SdkConfig.new(api_key: 'key').validate!
  failures << "FAIL: should raise for missing base_url"
rescue ArgumentError
  puts "  PASS: Raises ArgumentError for missing base_url"
end

begin
  ExperimentationPlatform::SdkConfig.new(base_url: 'http://example.com').validate!
  failures << "FAIL: should raise for missing api_key"
rescue ArgumentError
  puts "  PASS: Raises ArgumentError for missing api_key"
end

begin
  ExperimentationPlatform::SdkConfig.new(base_url: 'http://example.com', api_key: 'k').validate!
  puts "  PASS: Valid config does not raise"
rescue ArgumentError => e
  failures << "FAIL: Valid config raised: #{e.message}"
end

puts

# ---------------------------------------------------------------------------
# Test group 6: track never raises (no backend reachable)
# ---------------------------------------------------------------------------
puts ">>> Track Safety Tests"

client = ExperimentationPlatform::Client.new(base_url: 'http://127.0.0.1:9', api_key: 'k', timeout: 1)
begin
  sent = client.track('page_view', 'user-1')
  assert_true('track without a key and nothing cached sends nothing and returns true', sent == true, failures)
  sent = client.track('purchase', 'user-1', experiment_key: 'exp', value: 1.0)
  assert_true('track with a key against an unreachable host returns false (no raise)', sent == false, failures)
rescue StandardError => e
  failures << "FAIL: track raised #{e.class}: #{e.message}"
end

puts

# ---------------------------------------------------------------------------
# Summary
# ---------------------------------------------------------------------------
puts "=" * 60
if failures.empty?
  puts "All standalone tests PASSED"
  puts
  puts "Cross-SDK hash parity CONFIRMED:"
  puts "  hash_user(\"user-123\", \"my-flag\") = #{ExperimentationPlatform::FeatureFlagEvaluator.hash_user('user-123', 'my-flag')}"
  puts "  Expected:                          0.6927449859213084"
  puts "=" * 60
  exit 0
else
  puts "#{failures.length} test(s) FAILED:"
  failures.each { |f| puts "  - #{f}" }
  puts "=" * 60
  exit 1
end

File: sdk/php/test_standalone.php
<?php

declare(strict_types=1);

/**
 * Standalone test runner — no composer/autoloader required.
 *
 * Verifies the cross-SDK hash vector, the result types, the cache and the
 * client's request shapes (through a fake HttpClient) without any external
 * dependency or network. Useful on systems where composer is not set up.
 *
 * Run: php sdk/php/test_standalone.php
 */

foreach (
    [
        'src/Errors/ExperimentationException.php',
        'src/Errors/ApiException.php',
        'src/Errors/AuthException.php',
        'src/Errors/NetworkException.php',
        'src/SdkConfig.php',
        'src/HttpClient.php',
        'src/Cache.php',
        'src/FeatureFlagEvaluator.php',
        'src/FlagEvaluation.php',
        'src/Assignment.php',
        'src/BatchResult.php',
        'src/ExperimentationClient.php',
        'tests/FakeHttpClient.php',
    ] as $file
) {
    require_once __DIR__ . '/' . $file;
}

use ExperimentationPlatform\Assignment;
use ExperimentationPlatform\BatchResult;
use ExperimentationPlatform\Cache;
use ExperimentationPlatform\Errors\ApiException;
use ExperimentationPlatform\Errors\NetworkException;
use ExperimentationPlatform\ExperimentationClient;
use ExperimentationPlatform\FeatureFlagEvaluator;
use ExperimentationPlatform\FlagEvaluation;
use ExperimentationPlatform\SdkConfig;
use ExperimentationPlatform\Tests\FakeHttpClient;

$failures = [];
$passed   = 0;

function ok(string $name): void
{
    global $passed;
    $passed++;
    echo "  PASS  $name\n";
}

function fail(string $name, string $reason): void
{
    global $failures;
    $failures[] = "  FAIL  $name: $reason";
    echo "  FAIL  $name: $reason\n";
}

function check(string $name, bool $condition, string $reason = ''): void
{
    if ($condition) {
        ok($name);
    } else {
        fail($name, $reason !== '' ? $reason : 'assertion failed');
    }
}

function makeClient(FakeHttpClient $http, int $cacheTtl = 300): ExperimentationClient
{
    return new ExperimentationClient(
        new SdkConfig(baseUrl: 'https://api.example.com', apiKey: 'test-key', cacheTtl: $cacheTtl, timeout: 5),
        $http
    );
}

// =============================================================================
// Hash parity — cross-SDK compatibility vector (utility only)
// =============================================================================

echo "\n=== Hash Compatibility ===\n";

$result   = FeatureFlagEvaluator::hashUser('user-123', 'my-flag');
$expected = 0.6927449859213084;
check('hashUser("user-123", "my-flag") === 0.6927449859213084', abs($result - $expected) < 1e-10, "got $result");

$outOfBounds = false;
for ($i = 0; $i < 100; $i++) {
    $h = FeatureFlagEvaluator::hashUser("user-$i", "flag-$i");
    if ($h < 0.0 || $h >= 1.0) {
        $outOfBounds = true;
        break;
    }
}
check('all 100 hash values are in [0.0, 1.0)', !$outOfBounds);

$h1 = FeatureFlagEvaluator::hashUser('det-user', 'det-flag');
$h2 = FeatureFlagEvaluator::hashUser('det-user', 'det-flag');
check('hashUser is deterministic', $h1 === $h2);

$hEmpty = FeatureFlagEvaluator::hashUser('', '');
check('hashUser("", "") returns a float in [0, 1)', $hEmpty >= 0.0 && $hEmpty < 1.0);

$hUnicode = FeatureFlagEvaluator::hashUser('用户-123', '功能标志');
check('hashUser(unicode) returns a float in [0, 1)', $hUnicode >= 0.0 && $hUnicode < 1.0);

check('local evaluation was removed (the server decides)', !method_exists(FeatureFlagEvaluator::class, 'evaluate'));

// =============================================================================
// Result types
// =============================================================================

echo "\n=== Result Types ===\n";

$flag = new FlagEvaluation('dark-mode', true, ['theme' => 'dark']);
check('FlagEvaluation exposes key/enabled/config', $flag->key === 'dark-mode' && $flag->enabled && $flag->config === ['theme' => 'dark']);
check('FlagEvaluation::disabled() is off with a null config', !FlagEvaluation::disabled('x')->enabled && FlagEvaluation::disabled('x')->config === null);
check('FlagEvaluation::toArray()', $flag->toArray() === ['key' => 'dark-mode', 'enabled' => true, 'config' => ['theme' => 'dark']]);

$assignment = new Assignment('exp', 'v1', 'control', true, null);
check('Assignment exposes experimentKey/variantId/variantName/isControl/configuration',
    $assignment->experimentKey === 'exp' && $assignment->variantId === 'v1'
    && $assignment->variantName === 'control' && $assignment->isControl && $assignment->configuration === null);
check('Assignment::toArray() uses the API field names', $assignment->toArray()['variant_name'] === 'control');

$batch = new BatchResult(2, 0, null);
check('BatchResult::isOk() without failures', $batch->isOk());
check('BatchResult::isOk() with failures is false', !(new BatchResult(1, 1, [['error' => 'x']]))->isOk());

// =============================================================================
// Cache
// =============================================================================

echo "\n=== Cache ===\n";

$cache = new Cache(300, 3);
check('get() returns null for a missing key', $cache->get('x') === null);
$cache->set('x', 42);
check('get() returns the stored value', $cache->get('x') === 42);
$cache->clear();
check('clear() removes all entries', $cache->get('x') === null && $cache->count() === 0);

$cache->set('k0', 0);
$cache->set('k1', 1);
$cache->set('k2', 2);
$cache->set('k3', 3);
check('evicts the oldest entry at capacity', $cache->get('k0') === null && $cache->get('k3') === 3);

$cache->clear();
$cache->set('flag:2:u1:a', 'A');
$cache->set('flag:2:u2:b', 'B');
$cache->set('flag:2:u1:c', 'C');
check('valuesWithPrefix() lists one user\'s entries in insertion order', $cache->valuesWithPrefix('flag:2:u1:') === ['A', 'C']);

// =============================================================================
// Client — request shapes through the fake HttpClient
// =============================================================================

echo "\n=== Client (fake HTTP) ===\n";

$assignPath   = '/api/v1/tracking/assign';
$evaluatePath = '/api/v1/feature-flags/evaluate/dark-mode?user_id=user-123';
$trackPath    = '/api/v1/tracking/track';
$batchPath    = '/api/v1/tracking/batch';

$http = new FakeHttpClient();
$http->on('POST', $assignPath, [
    'experiment_key' => 'checkout-flow', 'user_id' => 'user-123', 'variant_id' => 'var-1',
    'variant_name' => 'treatment', 'is_control' => false, 'configuration' => ['color' => 'green'],
]);
$http->on('GET', $evaluatePath, ['key' => 'dark-mode', 'enabled' => true, 'config' => null]);
$http->on('POST', $trackPath, ['id' => 'evt-1']);
$http->on('POST', $batchPath, ['success_count' => 2, 'failure_count' => 0, 'errors' => null]);
$client = makeClient($http);

$a = $client->getAssignment('checkout-flow', 'user-123', ['country' => 'US']);
check('getAssignment() POSTs /api/v1/tracking/assign with experiment_key/user_id/context',
    $http->requests[0]['method'] === 'POST' && $http->requests[0]['path'] === $assignPath
    && $http->requests[0]['body'] === ['experiment_key' => 'checkout-flow', 'user_id' => 'user-123', 'context' => ['country' => 'US']]);
check('getAssignment() maps the response', $a instanceof Assignment && $a->variantName === 'treatment' && !$a->isControl && $a->configuration === ['color' => 'green']);
$client->getAssignment('checkout-flow', 'user-123');
check('getAssignment() is sticky (served from cache)', count($http->requestsTo($assignPath)) === 1);

$f = $client->evaluateFlag('dark-mode', 'user-123');
check('evaluateFlag() GETs /api/v1/feature-flags/evaluate/{key}?user_id=…', $http->requests[1]['method'] === 'GET' && $http->requests[1]['path'] === $evaluatePath);
check('evaluateFlag() maps the response', $f->key === 'dark-mode' && $f->enabled && $f->config === null);
check('isFeatureEnabled() is true (cache hit, no new request)', $client->isFeatureEnabled('dark-mode', 'user-123') && count($http->requestsTo($evaluatePath)) === 1);

$ok = $client->track('purchase', 'user-123', ['sku' => 'pro'], 'checkout-flow', null, 12.5);
$trackBody = $http->requestsTo($trackPath)[0]['body'] ?? [];
check('track() with a key POSTs one /api/v1/tracking/track', $ok && count($http->requestsTo($trackPath)) === 1);
check('track() body has event_type/event_name/user_id/experiment_key/value/metadata/timestamp',
    ($trackBody['event_type'] ?? null) === 'purchase' && ($trackBody['event_name'] ?? null) === 'purchase'
    && ($trackBody['user_id'] ?? null) === 'user-123' && ($trackBody['experiment_key'] ?? null) === 'checkout-flow'
    && ($trackBody['value'] ?? null) === 12.5 && ($trackBody['metadata'] ?? null) === ['sku' => 'pro']
    && is_string($trackBody['timestamp'] ?? null) && !array_key_exists('feature_flag_key', $trackBody));

$ok = $client->track('page_view', 'user-123', ['page' => '/']);
$batchEvents = $http->requestsTo($batchPath)[0]['body']['events'] ?? [];
check('track() without a key fans out one /api/v1/tracking/batch', $ok && count($http->requestsTo($batchPath)) === 1);
check('fan-out has one entry per cached assignment + evaluated flag',
    count($batchEvents) === 2 && ($batchEvents[0]['experiment_key'] ?? null) === 'checkout-flow'
    && ($batchEvents[1]['feature_flag_key'] ?? null) === 'dark-mode'
    && ($batchEvents[0]['metadata'] ?? null) === ['page' => '/']);

$before = count($http->requests);
check('track() without a key and nothing cached sends nothing', $client->track('page_view', 'user-999') && count($http->requests) === $before);

$r = $client->trackBatch([
    ['event_name' => 'a', 'user_id' => 'user-123', 'experiment_key' => 'checkout-flow'],
    ['event_name' => 'b', 'user_id' => 'user-123', 'feature_flag_key' => 'dark-mode'],
]);
check('trackBatch() returns a BatchResult mapped from the response', $r instanceof BatchResult && $r->successCount === 2 && $r->isOk());

$r = $client->trackBatch([['user_id' => 'u']]);
check('trackBatch() reports malformed events without sending them', $r->failureCount === 1 && ($r->errors[0]['index'] ?? null) === 0);

check('getAssignments()/getEvaluatedFlags() list the cache', count($client->getAssignments('user-123')) === 1 && $client->getEvaluatedFlags('user-123') === ['dark-mode']);

// Failures: never cached, never thrown
$http = new FakeHttpClient();
$http->on('POST', $assignPath, new ApiException('Experiment not found', 404));
$http->on('GET', $evaluatePath, new NetworkException('timeout'));
$http->on('POST', $trackPath, new RuntimeException('unexpected'));
$client = makeClient($http);

check('getAssignment() returns null on 404', $client->getAssignment('checkout-flow', 'user-123') === null);
$client->getAssignment('checkout-flow', 'user-123');
check('failed assignment is not cached (second request made)', count($http->requestsTo($assignPath)) === 2);
check('evaluateFlag() is disabled on a network error', !$client->evaluateFlag('dark-mode', 'user-123')->enabled);
try {
    $sent = $client->track('purchase', 'user-123', [], 'checkout-flow');
    check('track() never throws and returns false on error', $sent === false);
} catch (Throwable $e) {
    fail('track() never throws', 'threw ' . $e::class . ': ' . $e->getMessage());
}

// =============================================================================
// Summary
// =============================================================================

echo "\n" . str_repeat('=', 50) . "\n";
$total = $passed + count($failures);
echo "Results: $passed/$total tests passed\n";

if (empty($failures)) {
    echo "\nAll standalone tests PASSED\n";
    exit(0);
}

echo "\nFailed tests:\n";
foreach ($failures as $f) {
    echo "$f\n";
}
exit(1);
